### ISSUE TITLE
Show details about blocked trackers

### ISSUE THREAD

**Issue by pocmo** [mozilla-mobile/focus-android#594]
Dependencies:
- [ ] #1014

**Comment by pocmo** [comment]
nice to have (p2) and v1 (p1) doesn't make sense? :)

**Comment by ahunt** [comment]
Ooops! I'm still learning triage skills...

**Comment by pocmo** [comment]
@brampitoyo Do you already know whether this will require strings (+ accessibility strings)? I guess it will? We are close to string freeze.

**Comment by brampitoyo** [comment]
Yes. I think it will require some strings. At least those that say something like “[x] trackers blocked”.

Can you advise about the kind of data that would be most helpful to visualise?

By “most helpful”, I mean “visualisation that will help our users learn the most about how much Tracking Protection has helped them”.

So far, I’ve thought of only two things:

* trackers blocked on this page vs. trackers blocked since install
* requests blocked on this page vs. domains blocked on this page

Are there more? Should we pick one over the other? Ideally, I’d like to have only one visualisation.

Regardless, here are two ways that I thought we can do it:

![tracking-protection-menu-visualisation-i01](https://cloud.githubusercontent.com/assets/51007/26136075/df556a52-3b0d-11e7-81e7-cb6c22000e4b.png)

Thoughts?

**Comment by pocmo** [comment]
> Can you advise about the kind of data that would be most helpful to visualise?

That's a very good question and I do not really have a good answer for that. Currently we count the number of trackers blocked on the current site - I think that's a helpful number because it shows the user: We blocked something on this page and it gives the user a rough idea of how much trackers are used by the page -> "If something doesn't work correctly then it might be caused by all those blocked trackers".

@ahunt can comment more on how complex the implementation is, but we should be able to:
* Count number of trackers per page or session or total
* Count number of trackers for each category that we use in settings (ad trackers, analytics trackers, social trackers, content trackers)

> Regardless, here are two ways that I thought we can do it:

Visually I like the left option a lot. Although I do not fully understand what the "progress bar" means: From the mock it looks like it's the number of trackers on this page compared to all blocked trackers in total - so technically this bar would get smaller on every page load?

**Comment by pocmo** [comment]
In triage we decided to move this to the next iteration and not ship this in v1 (now that we are past string freeze).

**Comment by liuche** [comment]
I was talking to ahunt to see if this might be a good project for an intern.

@brampitoyo I'd be hesitant to include "total trackers blocked since install" because even in the mocks, that number seems overwhelmingly huge! One thing that I think could be more informational is listing information about the trackers that are being blocked. I'm not sure where or how this would fit in the UI (but I'm sure you would have ideas about this!), but I'd certainly be interested to know either the names or the domains or the urls of who's doing the tracking, that Focus is blocking from tracking the user.

For example, as a user, I can tell "how many trackers" are being blocked by Focus, but since it's just a number, it's not clear to me who is trying to track me! I would like more information about the specific trackers (is it Google, or ads, or something else?).

**Comment by brampitoyo** [comment]
@liuche You brought up a good point about total trackers blocked. All the number does is tell you about the scale of the problem, but it doesn’t tell you anything helpful about your situation or history.

Is 10,000 trackers blocked good? Is bigger number better, or does it mean that I’m browsing in a more insecure way – necessitating Focus to block more trackers? And why should I care about seeing any number at all?

Ok. So we’d like to add more information. But what kind? I think there are a lot of possibilities to explore.

Here’s a starter list!

#### Who are these trackers?

Categorise network requests by type. For example:

* Social networks (they track you across the web, right?): Facebook, Twitter, etc.
* Ad servers: AdSense, AdBrite, clicksor, and various ad networks.
* Web analytics: companies that do click tracking, for example.

Then when you tap on each category, we’ll tell you what domains are responsible.

#### What have these trackers done?

Add information that focuses on _how much better_ your browsing experience is.

* Minutes of time saved when loading the page
  * Provide context. For example: “You’ve saved a night’s sleep’s worth of time.”
* Megabytes of data saved when loading the page
  * Provide context. For example: “You’ve saved 25% of your monthly data plan.”
* Battery power saved when loading the page
  * Provide context. For example: “Your phone stays on 27 minutes longer.”

#### How can I avoid these trackers?

Add information that helps you actively make decisions on _places to frequent or avoid_.

* Top 10–20 websites in your history that has the most trackers
* Top 10–20 websites in your history that has the least (or no) trackers
* Top 10–20 tracking companies that Focus has blocked the most. Saying “doubleclick.net” isn’t helpful. Saying “Google” or “Amazon” is helpful.

What do you think?

**Comment by liuche** [comment]
I like these ideas, since the emphasis on **who**, **what**, **where** are clear questions people might have. It's also interesting to help improve people's mental model of what trackers and content blocking are actually doing (because I think that's not at all clear to most people).

What we show would probably depend on what information we can collect, so some initial thoughts:

* I like the idea of categorization! especially since we already allow people to block based on that in settings.

* I'm not sure we could collect the "amount of data savings" without actually downloading the content to see how big it is - but perhaps if you don't have certain things turned on, it could let you know how much data that's using. Not sure if that's a good or bad thing (e.g., will it feel like Firefox is pressuring you to block trackers by telling you how much data you're wasting?)

* Websites with the most trackers could be an interesting additional page to have (e.g. an overlay/dialog, or something in settings, or whatever) - I'm not sure that showing a users "top 10 websites" visited in Focus would be prudent though

Other ideas: there's an addon called Privacy Badger that shows a list of urls and what's being blocked on them. This might be more complicated by adding controls, rather than just read-only information, but maybe it's useful to see what those values are (e.g., is doubleclick.com useful information for most people?)

<img width="429" alt="screen shot 2017-06-02 at 13 16 02" src="https://cloud.githubusercontent.com/assets/1082253/26743660/8ab71ffe-4797-11e7-913c-47a08a1dd153.png">

**Comment by brampitoyo** [comment]
@pocmo What information can we collect in order to display better TP-related information? Are the things on my list viable to accomplish? I know that we’re using a third-party list powered by Disconnect, so we’d have to track some things on the user’s device. Would this slow performance down or increase site loading time?

**Comment by pocmo** [comment]
We need to refactor some of the blocking code to make this information available at runtime. But then we should be able to get counts for:

* Categories (the same we use in settings): Ad trackers, Analytics trackers, Social trackers, "Other content trackers" -> This shouldn't be too hard to implemented and what I really like about it is that it help the user to understand what those settings do.

* The full URL that was blocked (http://www.example.org/foo/bar/badsite) -> Not hard to implement but I do not really think that it's helpful to most users.

* The matched "domain" from the blocklist (example.org) -> Way easier to read than full URLs but maybe not that helpful either?

* Every "tracker group" has a label in the blocklist (which can be a company), e.g. "Google", "Facebook", "AdLeave", "Tumblr", .. -> This is quite interesting. And maybe regular users might be surprised by this ("Oh, Facebook tracks me on this random website?").


What we do not know is anything about time, traffic or battery saved. As we are only doing client side blocking this data is impossible to know without loading and/or processing the resource. And even if we load everything (because tracking is disabled for example) then we can't reliably get most of those measurements from WebView. -> With Gecko we might be able to ask for more detailed information about some loaded resources.

**Comment by ahunt** [comment]
As promised to @liuche , I'm finally adding some notes on how we could implement storing tracker "owners" or "groups" (i.e. Google/Facebook/AdLeave/Tumblr/etc) - aka the 4th bullet point from pocmo's previous comment - I've put my notes in issue #1014 to avoid putting too much clutter in this bug (especially since its unclear whether or not tracker owners are required).

**Comment by brampitoyo** [comment]
We’d like to:

* Remove details about the number of trackers blocked, as it was found to be inaccurate.
* Replace the string “Trackers blocked” with “Content Blocking”, to reflect the fact that we’re going to block more than trackers.
    * Blocking ads and (potentially) tracking cookies are planned for v8.0.
    * Blocking autoplay sound/video and HTTP pages may be planned in future versions.

<img width="360" src="https://user-images.githubusercontent.com/51007/45730932-dc90f000-bc28-11e8-8fdc-d26eea0eebee.png">

**Comment by vesta0** [comment]
@brampitoyo I agree that we should rename to "Content Blocking" but why remove # of trackers blocked - why isn't it accurate?

**Comment by brampitoyo** [comment]
> …why remove # of trackers blocked - why isn't it accurate?

@vesta0 I remember hearing from Engineering about how the # of trackers represented a “best guess”, rather than being a figure that’s accurate every single time you load any website. It could’ve been because we use WebView, and they can’t give us an accurate figure?

Either way, my contention with using numbers without citing any context is that it’s impossible to know what “good” looks like.

For example, if the figure shows a “0”, it shows that the site doesn’t have any trackers, right? But a zero figure looks bad. The user may have the impression that Content Blocking isn’t doing its job.

On the other hand, if the figure shows a number, then how can I link it to a sense that “I’m fully secure” vs. “I’m potentially insecure”? Is larger number better than smaller numbers? In most places, users are accustomed to the fact that larger numbers are almost always better. But in Firefox Focus, the reverse is true. Smaller is better. Although smaller also implies that Firefox Focus is doing less blocking.

The problem of using raw numbers without any context is so tricky, that I recommended doing away with numbers entirely, and just having a toggle that says whether Content Blocking is on or off. If it’s on, it’s always on whatever the number may be.

-----

In my post above – https://github.com/mozilla-mobile/focus-android/issues/594#issuecomment-305375821 – I’ve written about a few things that Firefox Focus can show, that will help give users some perspectives about things that are blocked on the page.

After looking at it again, I have a different proposal.

Instead of listing the name of companies, megabytes of data saved, etc., we can list _categories of element_ that Content Blocking is currently blocking on the page,.

Right now, our main menu shows this at all times:

> **Content Blocking** [toggle ON]
> Turn off to fix some sites

We can revise it, so that when no blockable element is detected, it shows the same as above:

> **Content Blocking** [toggle ON]
> Turn off to fix some sites

But when things are blocked, we list _categories of elements_ without quantifying them. Like this:

> **Content Blocking** [toggle ON]
> Ads, trackers

Or this:

> **Content Blocking** [toggle ON]
> Cookies

What we show under the “Content Blocking” header depends on what elements are currently blocked.

This gives users the idea of how hard Firefox Focus is currently working to protect them (where showing more items on the list means more blocking), without showing any quantity (where a bigger number is actually not better for the user’s security situation).

**Comment by brampitoyo** [comment]
On #3841, we not only decided to keep the number of trackers. We’ve also made that number more visible by including the shield icon in the URL bar.

Doing that should resolve this issue.
